snippet_parser/core_test.py

```diff
@@ -146,5 +146,63 @@ def test_no_wikicode_in_section_titles(self):
             "== Section title ''with Wikicode'' == \n\nIrrelevant content{{cn}}")
         self.assertEqual(section, 'Section title with Wikicode')
 
+    def _do_extract_sentences(self, wikitext, sentences):
+        for sentence in sentences:
+            start = wikitext.find(sentence)
+            if start < 0: continue
+            marked = core._SENTENCE_MARKER_MARKUP.format(sent = sentence)
+            wikitext = wikitext[:start] + marked + wikitext[start+len(sentence):]
```

Comment:
Why do we do this here if `extract_from_sentences` will also do it for us when we call it below? Remember that parse_live, the real caller of the code we're testing, does not need to mark the sentences in advance, so it's suspicious that we're doing it in the test...

Reply:
Here it's like in `_do_extract` you replace the {citation_needed_tmpl} with the citation needed marker so you can set up the fake value with mocks.
```
        html = html.format(
            citation_needed_tmpl = core._CITATION_NEEDED_MARKER_MARKUP)
        html = html.format(tpl = cn_expansion)
        self._set_wikipedia_parse_response(html)
```
If I didn't do this beforehand, the wikitext wouldn't have sentence markers, then we couldn't set up an expected fake value with mocks.

Reply:
Oh, yeah, but `_do_extract` does that to set up the fake response from the Wikipedia API (the `html` variable), but does not pass that to `extract` -- what we pass to `extract` is the unmodified `wikitext` argument.

So the arguments to `_do_extract` are:

- Fake HTML containing the string `{citation_needed_tmpl}`, which is we'll replace with the actual citation needed marker. Ultimately, we pretend the MediaWiki API returned this HTML.
- Wikitext to use for testing. This is just passed unmodified to `extract`, which will do its processing, then call the Wikipedia API and receive the fake HTML, then do more processing on that.
- Some fake text for the text contents of the HTML expansion of the `{{ cn }}` template. This argument doesn't really matter much, I don't even remember if it's useful offhand.

Here it seems like you're building the fake HTML response but assigning it to the `wikitext` variable, and then both using that as the fake HTML and passing it in to `extract_from_sentences`.

Reply:
Yeah, right. My code does make people confused.
How about I use a `html` variable to set up the fake response, and pass unmodified wikitext to `extract_from_sentences`? The wikitext is optional that only if we want to test the modified wikitext, we pass wikitext to `_do_extract_sentences`.
```
def _do_extract_sentences(self, html, sentences, wikitext=''):
    markers = []
    for sent in sentences:
        marker = core._SENTENCE_MARKER_MARKUP.format(sent = sent)
        markers.append(marker)
    html = html.format(*markers)
    self._set_wikipedia_parse_response(html)
    ret = self._sp.extract_from_sentences(wikitext, sentences)
    call_args = None
    if self._wp.parse.call_count:
        call_args = self._wp.parse.call_args[0][0]
    ret = ret[0] if ret else []
    return call_args, ret
```
A test case is like: 
```
def test_sentence_to_snippet(self):
    sentences = ['A sentence.']
    html = '<p>Stuff. {0} More stuff.</p>'
    _, [_, snippets]  = self._do_extract_sentences(html, sentences)
    expected = '<div class="%s"><p>Stuff. ' % (core.SNIPPET_WRAPPER_CLASS
        ) + core._SENTENCE_MARKER_MARKUP.format(sent = 'A sentence.'
        ) + ' More stuff.</p></div>'
    self.assertEqual(expected, snippets[0])
```

Reply:
Yep, I think that makes sense. It might make sense to actually pass and check the wikitext in all tests I guess (including the existing ones), since otherwise we're only really exercising the code that manipulates the HTML. But this is also fine as it is.

Reply:
Thanks! Yeah, I understand what you mean. :)